standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/parser/Filter.g

```diff
@@ -141,7 +141,11 @@ binOpExpression
        (
          (key = Identifier op = operator  value = DateLiteral)
          |
+         (key = Identifier op = operator keyword = KW_DATE value = StringLiteral)
+         |
```

Comment:
Seems that DateLiteral supports optional `date` keyword prefix but does not support quotes:
https://github.com/apache/hive/blob/2fb969c6c6c7ebc9d2c34e27152c147aa34d7ff9/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/parser/Filter.g#L464-L467

How about improve `DateString` by adding a quoted alternative
```
fragment DateString
    :
    '\'' (Digit)(Digit)(Digit)(Digit) '-' (Digit)(Digit) '-' (Digit)(Digit) '\''
    | (Digit)(Digit)(Digit)(Digit) '-' (Digit)(Digit) '-' (Digit)(Digit)
    ;
```
instead of adding these alternatives to all occurrences where `DateLiteral` is referenced.
Btw I don't know why `DateString` requires quotes in the first place since `StringLiteral` has

Reply:
Yes. That would make more sense.
Thanks for pointing out.

Reply:
After i did the changes. I wonder the following test case 

`create table test_tbl (a int) partitioned by (b string);
insert into table test_tbl partition(b='2020-01-01') values (1);
insert into table test_tbl partition(b='2020-02-01') values (1);
insert into table test_tbl partition(b='2020-03-01') values (1);

select * from test_tbl where b = '2020-03-01';`

The string data type would match the dateString which will lead to result inconsistency.

Reply:
@kasakrisz Consider the following case where string column value matches the pattern of the date

create table test_date (a int) partitioned by (b string);
insert into table test_date partition(b='2020-20-20') values (1);
insert into table test_date partition(b='2015-08-09') values (1);
insert into table test_date partition(b='2016-08-09') values (1);

select * from test_date where date_c='2020-20-20';

Here '2020-20-20' would match the dateLiteral and hence will lead to inconsistent result right? I am missing something?

Reply:
Got your point.

How about this?
```
fragment DateString
    :
    (Digit)(Digit)(Digit)(Digit) '-' (Digit)(Digit) '-' (Digit)(Digit)
    ;
DateLiteral
    :
    KW_DATE '\'' DateString '\'' { ExtractDate(getText()) != null }?
    | DateString { ExtractDate(getText()) != null }?
    ;
```
* If `DATE` keyword presents let's expect a valid date in quotes.
* If `DATE` keyword does not present treat it as an unquoted date. I haven't been able to generate a filter having a date literal represented like this so I just added this to be backward compatible with the current solution.
* When cbo is off `DATE` keyword is not added to the filter string and the date is in quotes. In this case date is treated as a string.

Reply:
It makes sense now.
I have changed the code. The test have passed now.

@kasakrisz Could you please review  the changes?